### ISSUE TITLE
The contents shown at the beginning of sections should only be activated if needed

### ISSUE THREAD

**Issue by tthuem** [SoftVarE-Group/SlideTemplate#28]
This is a useful feature for lectures, but certainly not for a normal presentation. Instead, it could be useful to have a slide showing only the topic of the next part and as the default no slide at all.

In theory, it would be feasible to not have any sections at all (to avoid the extra slides), but then there is also no further structure in the PDF content overview.

**Comment by Benno2000** [comment]
I added two booleans `sectiontitleslide` and `sectionoverview` to enable the section title slide with an optional section overview. These are the results:

With `\sectiontitleslidetrue` only:

![Bildschirmfoto 2022-08-02 um 15 20 40](https://user-images.githubusercontent.com/23581917/182386272-ad2a3472-8ed9-460a-b023-edbb8674f28f.jpg)

With `\sectiontitleslidetrue` and `\sectionoverviewtrue`:

![Bildschirmfoto 2022-08-02 um 15 20 26](https://user-images.githubusercontent.com/23581917/182386421-a089d360-3b26-46f3-a9b1-3cb2c7b1cd7b.jpg)

The default (if `sectiontitleslide` is not set to true) is no automatic slide at the beginning of a section.

**Comment by EagleoutIce** [comment]
This could be made into package options via 

```latex
\DeclareOption{sectiontitleslides}{\sectiontitleslidestrue}
% ....
\ProcessOptions
```

What do you say?

**Comment by Benno2000** [comment]
You're right, that looks much more convenient to use! I also wonder if i should use `sectionoverviews` as a single option instead of needing to use both options to get a slide with title + table of contents. In my opinion that would make more sense from a user's point of view, what do you think? 
Therefore we could maybe also use an option that accepts different values instead (e. g. `sectionstart=sectionoverviews`). Do you know how something like that is possible?

**Comment by EagleoutIce** [comment]
In theory we could make `sectionoverviews` set both flags to true and use something like `nosectiontitleslide` to disable the title slides.

Jeah free value keys are possible, there are several packages implementing that behavior (see this [stackexchange](https://tex.stackexchange.com/questions/417704/key-value-syntax-in-package-options)). If nothing too fancy is required, [kvoptions](https://www.ctan.org/pkg/kvoptions) should suffice. If you prefer the pgfkeys interface, we could rely on [pgfopts](https://www.ctan.org/pkg/pgfopts). In the following i explain, how i would implement a generic way of achieving the behavior (while presenting some alternatives). The approach can be made much simpler if:
1. You want only a small, fixed set of allowed values
2. The user should not be able to create own definitions

I for myself think, that the option with `nosectiontitleslide` should suffice.

With kvoptions, the Code would look roughly like this:
```latex
\RequirePackage{kvoptions}

\SetupKeyvalOptions{family=uulm, prefix=uulm@}
\DeclareStringOption{sectionstart}
% ...
\ProcessKeyvalOptions*
```

Similarly, if you do want want to use another package, you can use beamers way of `\DeclareOptionBeamer` roughly like this (i have no time to test the code at the moment):

```latex
\DeclareOptionBeamer{sectionstart}[sectionoverviews]{\def\@uulm@sectionstart{#1}}
% ...
\ProcessOptionsBeamer
```
Independent from the option processing, we move on.

Now we may want to act according to the selected option. I would make a macro for each "section start" definition (if you prefer a predefined set) like:

```latex
\newcommand{\uulm@sectionstart@sectionoverviews}{%
% ... frame definition
}
```

In the preamble (after `\ProcessKeyvalOptions` and after all those macro-definitions are in-place[^1]). To enable the respective drawer we use another macro:

```latex
\newcommand{\uulm@process@selectionstart}{
   \ifcsname uulm@sectionstart@\@uulm@sectionstart\encsname
      % the drawer requested exists! Set it
      \letcs\uulm@sectionstart@do{@uulm@sectionstart@\uulm@sectionstart}
  \else % the drawer does not exist!
      \PackageError{beamerthemeuulm}{'\@uulm@sectionstart' is not a valid sectionstart pick one of '... <TODO>'}{}
      \endinput % or whatever behavior is desired
  \fi
}
% \AtEndPreamble{
   \uulm@process@selectionstart
% }
```

With this in place we can use `\uulm@sectionstart@do` wherever we want the section starter to be placed (actually, this is an indirection we can avoid by directly inserting the frames with `\AtBeginSection` - i would make this up to you):

```latex
\AtBeginSection{\uulm@sectionstart@do}
```


[^1]: If you want the user to define his/her own drawers or if you want to make sure, that all drawers are respected, you can use [etoolbox'](https://www.ctan.org/pkg/etoolbox) `\AtEndPreamble`-macro for the following code - it should be already active with beamer.

**Comment by Benno2000** [comment]
Thanks for the detailed explanation! However, i couldn't get it working yet but one of the goals when creating the new template also was to keep it simple and therefore understandable for others. 
So for the sake of simplicity i would also use the option where`\sectionoverviews` simply sets both flags to true. Would we even need something like `nosectiontitleslide` as this is the default behaviour when none of the options is used?

**Comment by EagleoutIce** [comment]
Jeah, that is the reason i added comments to my other PRs. However, if you want to get it working regardless of that, i could take a look,

Regarding the negative flag(s), i am always a fan of specifying defaults explicitly, therefore, I would want to write that i do not want the feature 'X' regardless of if that would be the default or not (how should i, as a user, know, what the default is and if it might be subject to change?). However, if we only specify positive flags end ensure backwards compatibility, there would be no "need" for these negative option.

**Comment by Benno2000** [comment]
In my latest commit (https://github.com/SoftVarE-Group/SlideTemplate/commit/60a5cefe178899b74c60e89268af8eef183e64d5) I added a `nosectiontitleslides` option and set the automatic section overviews as default for backwards compatibility.

**Comment by EagleoutIce** [comment]
Setting the flags directly is fine (by using `\...true`) if you want to make defaults more explicit you could use `\ExecuteOptions{...}`.

**Comment by Benno2000** [comment]
Thanks for the hint, I changed it in my latest commit (https://github.com/SoftVarE-Group/SlideTemplate/commit/3a7a4689d9b3f3a88f9b2f93bdee9b0d277454a6).

**Comment by EagleoutIce** [comment]
Jeah that works for me (but isn't it "enabled *by* default"?),
I think, the branch is ready for a pull request?

**Comment by Benno2000** [comment]
I think both options work, but i found that you rather use "by" in software engineering :) (I also used it in the demo file).

I'll change it and then create a pull request. I'm just not sure whether we should merge into the main or release branch (@tthuem told me to work on the release a while ago, but if I've gotten it right, you are branching from main?).

**Comment by EagleoutIce** [comment]
I was not aware of this. Yes, I did branch from main but I can change that if you prefer the release branch.

**Comment by Benno2000** [comment]
> I was not aware of this. Yes, I did branch from main but I can change that if you prefer the release branch.

I asked @tthuem again, we should work on the release branch. So it would be good if you could change your PRs to that branch :)

**Comment by EagleoutIce** [comment]
Done :)

**Comment by Benno2000** [comment]
Alright, then I'll merge my changes into release.

**Comment by tthuem** [comment]
I would propose that the default is to have no section and title overviews, as the template is typically used for presentations. The other two can then be activated if needed.

**Comment by Benno2000** [comment]
I changed this in https://github.com/SoftVarE-Group/SlideTemplate/commit/ef09a80242fa94544c304a4f2e445c27a5516834 and made no section overviews the default behaviour. Do you think we still need the option `nosectionoverviews` then?

**Comment by tthuem** [comment]
No, there is no need for this flag then.

**Comment by Benno2000** [comment]
Alright, I removed the flag in b13e99339ca14835e1f21a674a0e8c71df797d68. If there is nothing else to do on this issue, I'll create a pull request.

**Comment by EagleoutIce** [comment]
Don't you want to use https://github.com/SoftVarE-Group/SlideTemplate/pull/35?

**Comment by Benno2000** [comment]
You are right, we already have a pull request for this one 😄

**Comment by tthuem** [comment]
The new parameters are still missing in our demo slides.

**Comment by tthuem** [comment]
I have thought about this issue again. I think the following would be great:

Default without handout mode: sectiontitleslides
Default with handout mode: sectionoverviews

And then three parameters to overwrite the default behavior: nosectionframes, sectiontitleslides, sectionoverviews

I actually like these title slides during the presentation, as the force you to stop for a moment and introduce what you talk about next. However, for handouts, these slides are rather useless, as one cannot navigate with them. During the presentation, already showing the subsections is typically not very helpful. The above default behavior would also ensure that page numbers with and without handout mode are identical, which could avoid confusion.

What do you think?

**Comment by Benno2000** [comment]
> The new parameters are still missing in our demo slides.

I don't know what you mean, I added the parameters to the `\documentclass`-command in the demo slides:
```
\documentclass[
	aspectratio=169, % default is 43
	8pt, % font size, default is 11pt
	%sectiontitleslides, % enable an automatic section title slide at the begin of each section
	sectionoverviews, % enable an automatic section overview at the begin of each section
	%handout, % handout mode without animations, comment out to add animations
]{beamer}
```

> Default without handout mode: sectiontitleslides
Default with handout mode: sectionoverviews

I like this idea, as if you want any other behavior, you can still overwrite the defaults. I'll try to implement it.

**Comment by Benno2000** [comment]
I implemented this in eb9a754ddf6ad43e6a5621bbe4e16c6acc5cb484.

**Comment by Benno2000** [comment]
I merged this into release.
